### ISSUE TITLE
Treat all warnings as errors

### ISSUE THREAD

**Issue by AnastaZIuk** [Devsh-Graphics-Programming/Nabla#685]
Nabla should be compiled with `/WX` flag enabled for [compiler](https://learn.microsoft.com/en-us/cpp/build/reference/compiler-option-warning-level?view=msvc-170) and [linker](https://learn.microsoft.com/en-us/cpp/build/reference/wx-treat-linker-warnings-as-errors?view=msvc-170) which treats warning as errors, then all of them should be eliminated and sources adjusted. Doing so may save us in future, recursive warning generation may lead to unresponsive builds taking more then half an hour, also we decrease a chance for undefined behaviours.

**Comment by AnastaZIuk** [comment]
this should be toggle-able, introduce `NBL_TREAT_WARNINGS_AS_ERRORS` and enabled by default but let user to override it as CMake option

**Comment by devshgraphicsprogramming** [comment]
You'll run into problems with that, because whenever you include a 3rdparty header you get all of its warnings.

So it's not as easy as just enabling WX because you don't want to be bogged down into forking every dependency and "fixing" their warnings.

P.s. there was a story one about an intern or a junior who fixed uninitialised variables in openssl and broke it cause they were actually a source of entropy for the rand gen

**Comment by AnastaZIuk** [comment]
I would limit fixing them to our headers & sources only

**Comment by AnastaZIuk** [comment]
> who fixed uninitialised variables in openssl and broke it cause they were actually a source of entropy for the rand gen

yeah I get you, though I don't think that would be a case if this happened in our codebase, I rather stick to initialize variables not to some random values which may lead to bugs, but of course it's not a case as well to fix all 3rdparty headers

**Comment by AnastaZIuk** [comment]
> You'll run into problems with that, because whenever you include a 3rdparty header you get all of its warnings.

well another problem this could introduce is an implicit vendor/toolset dependency, but since we are targeting latest toolsets and cpp standard I would not care much

**Comment by AnastaZIuk** [comment]
by the way, there are solutions how to disable/ignore warnings in external headers coming from 3rdparty and still treat all warnings as errors with our codebase

```
#pragma warning(push, 0)
// Some include(s) with unfixable warnings
#pragma warning(pop)
```

also https://devblogs.microsoft.com/cppblog/broken-warnings-theory/

**Comment by AnastaZIuk** [comment]
> by the way, there are solutions how to disable/ignore warnings in external headers coming from 3rdparty and still treat all warnings as errors with our codebase
> 
> ```
> #pragma warning(push, 0)
> // Some include(s) with unfixable warnings
> #pragma warning(pop)
> ```
> 
> also https://devblogs.microsoft.com/cppblog/broken-warnings-theory/

with that you don't have this problem you mentioned

> because whenever you include a 3rdparty header you get all of its warnings.
